File: bench/benches/hash.clj
;; Mutable hash-table throughput (see bench/benches/hash.l). Clojure's core maps
;; are persistent, so this uses java.util.HashMap for a true mutable table.
(load-file "lib/bench.clj")
(defn hash-run [n]
  (let [h (java.util.HashMap.)]
    (dotimes [i n] (.put h (+ (* 97 i) 1) i))
    (let [scan (fn [] (loop [i 0 a 0]
                        (if (< i n) (recur (inc i) (+ a (.get h (+ (* 97 i) 1)))) a)))
          a (scan)]
      (dotimes [i n]
        (let [k (+ (* 97 i) 1)]
          (.put h k (+ (.get h k) 1))))
      (+ a (scan)))))
(bench "hash" (fn [] (hash-run 10000)))

File: bench/benches/hash.rs
include!("../lib/bench.rs");

use std::collections::HashMap;

// mutable hash-table throughput (see bench/benches/hash.l). checksum = N*N.
const N: i64 = 10000;

fn main() {
    bench("hash", || {
        let mut h: HashMap<i64, i64> = HashMap::new();
        for i in 0..N {
            h.insert(97 * i + 1, i);
        }
        let mut a: i64 = 0;
        for i in 0..N {
            a += h[&(97 * i + 1)];
        }
        for i in 0..N {
            let k = 97 * i + 1;
            let v = h[&k] + 1;
            h.insert(k, v);
        }
        for i in 0..N {
            a += h[&(97 * i + 1)];
        }
        a
    });
}
